### ISSUE TITLE
Rack 3.1 is missing content-length on 0 byte responses

### ISSUE THREAD

**Issue by zarqman** [rack/rack#2218]
In a Rails 7.1 app, empty responses are now missing the `content-length` header when used with Rack 3.1. On Rack 3.0, the header is present with the expected value of `0`.

I'm not positive, but this might be a byproduct of #2149 since that deferred setting `content-length`. If there's no body to iterate, perhaps `content-length` gets missed?

I don't have a thorough grasp on how Rails and Rack interact in the case of an empty response, so it's possible this issue belongs to Rails instead of Rack. However, I'm starting here since changing the Rack version alone is enough to trigger the behavior difference.

##### Why this matters

Some HTTP servers revert to chunked encoding when the payload size is unknown. This is inefficient.

Additionally, some HTTP caches use `content-length` to decide what is and isn't cacheable and its absence can cause the entire request to become uncacheable, causing excess load on app servers.

<details>
<summary>Reproduction example</summary>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails"
  # gem "rack", "~> 3.0.0"
  gem "rack", "~> 3.1.0"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  config.hosts << "www.example.com"
  config.secret_key_base = "secret_key_base"
  config.action_dispatch.show_exceptions = :rescuable
  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/head" => "test#one"
    get "/empty" => "test#two"
    get "/body" => "test#three"
  end
end

class TestController < ActionController::Base
  def one
    head 200
  end
  def two
    render plain: ''
  end
  def three
    render plain: 'a', layout: false
  end
end

require "minitest/autorun"

class TestControllerTest < ActionDispatch::IntegrationTest

  # works on rack 3.0
  # content-length key is missing on rack 3.1
  def test_head
    get "/head"
    assert_response 200
    assert_equal '0', headers['content-length']
    # assert headers.key?('content-length')
  end

  # works on rack 3.0
  # content-length key is missing on rack 3.1
  def test_empty_body
    get "/empty"
    assert_response 200
    assert_equal '0', headers['content-length']
    # assert headers.key?('content-length')
  end

  # works on rack 3.0 and 3.1
  def test_with_body
    get "/body"
    assert_response 200
    assert_equal '1', headers['content-length']
  end

  private
    def app
      Rails.application
    end
end
```

</details>

**Comment by dentarg** [comment]
Wow, I was just now taking a look at https://github.com/sinatra/sinatra/issues/2019 😅 

My observation (from Sinatra tests) is that it does not happen with rack head, what more has changed?

**Comment by zarqman** [comment]
FWIW, I just tested my repro example above against rack head and the issue still occurs, same as with Rack 3.1.

**Comment by dentarg** [comment]
You are right. That was incorrect info from me. Sinatra [can't even start its tests](https://github.com/sinatra/sinatra/actions/runs/9565078027/job/26367227454#step:5:13) at rack head currently.

**Comment by Earlopain** [comment]
Yeah, the PR you found seems to be the cause of this. It prevents a content-length of 0 here: https://github.com/rack/rack/blob/800e53fbe15b3424b7a8946b067bf6f2e648d5a8/lib/rack/response.rb#L113-L113

Removing that condition makes your test pass but introduces a few other failures where I think middlewares are part of the stack that would too eagerly set this header to 0 before something was written.

https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6

> Any Content-Length field value greater than or equal to zero is valid

**Comment by jeremyevans** [comment]
@ioquatix I think we should revert #2149.  If you can come up with an alternative approach that is backwards compatible, we can consider that.

Regarding failures showing up as successes, that's probably an issue in bake-test-external that needs to be fixed.   I mentioned it in https://github.com/ioquatix/bake-test-external/issues/6, but probably should have created a separate issue for it.

**Comment by ioquatix** [comment]
I will need to investigate before I can advise on the right course of action. However, I'm not convinced there is a problem yet. I'll need to see it for myself.

**Comment by ioquatix** [comment]
I experimented with this code and found a number of bugs and issues, some in Rack and some in Sinatra.

In particular, `MockResponse#initialize` invokes `buffered_body!` which was previously adding a `content-length` header even if it did not exist in the original response, potentially tricking tests from knowing the actual response headers.

In a follow up PR, I'd like to consider adding a flag to `MockResponse` which does not call `buffered_body!` as some tests DO want to see the original body (chunks, streaming, etc).

**Comment by ioquatix** [comment]
I've also confirmed, with the changes proposed in #2219, the OP's test cases now pass.

**Comment by ioquatix** [comment]
This should be fixed in Rack v3.1.7.

**Comment by ioquatix** [comment]
```
> ruby ./test.rb
Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
Fetching rack 3.1.7
Installing rack 3.1.7
> ruby ./test.rb
Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
Fetching rack 3.1.7
Installing rack 3.1.7
Run options: --seed 46342

# Running:

I, [2024-07-11T13:48:44.219721 #428134]  INFO -- : -----------------------------
I, [2024-07-11T13:48:44.219748 #428134]  INFO -- : TestControllerTest: test_head
I, [2024-07-11T13:48:44.219753 #428134]  INFO -- : -----------------------------
I, [2024-07-11T13:48:44.231784 #428134]  INFO -- : Started GET "/head" for 127.0.0.1 at 2024-07-11 13:48:44 +1200
.I, [2024-07-11T13:48:44.275496 #428134]  INFO -- : ----------------------------------
I, [2024-07-11T13:48:44.275512 #428134]  INFO -- : TestControllerTest: test_with_body
I, [2024-07-11T13:48:44.275518 #428134]  INFO -- : ----------------------------------
I, [2024-07-11T13:48:44.275784 #428134]  INFO -- : Started GET "/body" for 127.0.0.1 at 2024-07-11 13:48:44 +1200
.I, [2024-07-11T13:48:44.279342 #428134]  INFO -- : -----------------------------------
I, [2024-07-11T13:48:44.279353 #428134]  INFO -- : TestControllerTest: test_empty_body
I, [2024-07-11T13:48:44.279358 #428134]  INFO -- : -----------------------------------
I, [2024-07-11T13:48:44.279552 #428134]  INFO -- : Started GET "/empty" for 127.0.0.1 at 2024-07-11 13:48:44 +1200
.

Finished in 0.062150s, 48.2704 runs/s, 96.5408 assertions/s.
3 runs, 6 assertions, 0 failures, 0 errors, 0 skips
```
